# This is my 3rd Project. 

In this one I will work with *dataset of used cars from eBay Kleinanzeigen*, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle.

The aim of this project is to **clean the data and analyze the included used car listings**. I will follow the instructions from **DataQuest** and pratice what I've learned so far.

The data dictionary provided with data is as follows:

* *dateCrawled* - When this ad was first crawled. All field-values are taken from this date.
* *name* - Name of the car.
* *seller* - Whether the seller is private or a dealer.
* *offerType* - The type of listing
* *price* - The price on the ad to sell the car.
* *abtest* - Whether the listing is included in an A/B test.
* *vehicleType* - The vehicle Type.
* *yearOfRegistration* - The year in which the car was first registered.
* *gearbox* - The transmission type.
* *powerPS* - The power of the car in PS.
* *model* - The car model name.
* *kilometer* - How many kilometers the car has driven.
* *monthOfRegistration* - The month in which the car was first registered.
* *fuelType* - What type of fuel the car uses.
* *brand* - The brand of the car.
* *notRepairedDamage* - If the car has a damage which is not yet repaired.
* *dateCreated* - The date on which the eBay listing was created.
* *nrOfPictures* - The number of pictures in the ad.
* *postalCode* - The postal code for the location of the vehicle.
* *lastSeenOnline* - When the crawler saw this ad last online.

Firstly we will start importing library and reading dataset into pandas and numpy.

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos.head(3)

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
dateCrawled            371528 non-null object
name                   371528 non-null object
seller                 371528 non-null object
offerType              371528 non-null object
price                  371528 non-null int64
abtest                 371528 non-null object
vehicleType            333659 non-null object
yearOfRegistration     371528 non-null int64
gearbox                351319 non-null object
powerPS                371528 non-null int64
model                  351044 non-null object
kilometer              371528 non-null int64
monthOfRegistration    371528 non-null int64
fuelType               338142 non-null object
brand                  371528 non-null object
notRepairedDamage      299468 non-null object
dateCreated            371528 non-null object
nrOfPictures           371528 non-null int64
postalCode             371528 non-null int64
lastSeen              

From the work I did in the last screen, I can make the following observations:

* The dataset contains 20 columns, most of which are strings.
* Most of the columns have non-null values.
* The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

Let's try to edit the columns' names to be easily readable.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  registration_year    gearbox  power_ps  model  \
0    480    test          NaN               1993    manuell         0   golf   
1  18300    test        coupe               2011    manuell       190    NaN   
2   9800    test          suv               2004  automatik       163  grand   
3   1500    test   kleinwagen               2001    manuell        75   golf   
4   3600    test   kleinwagen               2008    manuell        69  fabia   

   odometer  registration_month fuel_type       brand unrepaired_damage  \
0    150000                   0    benzin  volkswagen               NaN   
1    125000                   5    diesel        audi                ja   
2    125000                   8    diesel        jeep               NaN   
3    150000                   6    benzin  volkswagen              nein   
4     90000                   7    diesel       skoda              nein   

            ad_created  num_photos  postal_code            last_seen  
0  2016-03-24 00:00:00           0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00           0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00           0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00           0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00           0        60437  2016-04-06 10:17:21

The columns' name is now changed into snakecase, which is easier to read and access.

Now we will explore some basic data to determine what to be done.

Initially we will look for: 
* Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
* Examples of numeric data stored as text which can be cleaned and converted.

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model       odometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count             299468               371528    371528.0  371528.00000   
unique                 2                  114         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq              263182                14450         NaN           NaN   
mean                 NaN                  NaN         0.0   50820.66764   
std                  NaN                  NaN         0.0   25799.08247   
min                  NaN                  NaN         0.0    1067.00000   
25%                  NaN                  NaN         0.0   30459.00000   
50%                  NaN                  NaN         0.0   49610.00000   
75%                  NaN                  NaN         0.0   71546.00000   
max                  NaN                  NaN         0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

There are text comlumns, which like *seller, offer_type, etc* has only 2 types.

Futhermore, *num_photos* has almost everything 0. And *price*, *odometer* are numeric values but stored as texts.

We will look closer into them and edit again.

In [7]:
autos['price'].value_counts()

0          10778
500         5670
1500        5394
1000        4649
1200        4594
2500        4438
600         3819
3500        3792
800         3784
2000        3432
999         3364
750         3203
650         3150
4500        3053
850         2946
2200        2936
700         2936
1800        2886
900         2874
950         2793
1100        2772
1300        2757
300         2731
3000        2720
550         2591
1600        2570
5500        2543
350         2514
400         2442
1250        2441
           ...  
22599          1
4782           1
8008           1
20290          1
50100          1
3245           1
58800          1
2795000        1
75400          1
10830          1
588            1
31555          1
3755           1
19269          1
14666          1
28995          1
4175           1
10670          1
77             1
8265           1
145900         1
2730           1
12973          1
42840          1
30530          1
23456          1
171000         1
21830         

In [8]:
autos['odometer'].value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: odometer, dtype: int64

We will rename 'odometer' and add in 'km' to be readable.

In [9]:
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

In [10]:
print(autos['price'].unique().shape)
autos['price'].describe()

(5597,)


count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [11]:
print(autos['price'].value_counts().sort_index(ascending=False).head(10))
autos['price'].value_counts().sort_index(ascending=True).head(20)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
Name: price, dtype: int64


0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: price, dtype: int64

We see 1,421 result '0' so it will be removed, however since eBay is popular with bidding, so '1' is resonable and we will keep it until '350,000', which is good enough. 

Anything higher than that should be considered 'fraud', because you don't bidd something that worth around $500,000 in internet.

In [12]:
autos = autos[autos['price'].between(1,351000)]

Now we will explore the date columns, which are 5 of them:
* `date_crawled`: added by the crawler
* `last_seen`: added by the crawler
* `ad_created`: from the website
* `registration_month`: from the website
* `registration_year`: from the website

Right now, `the date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like `Series.describe()` to understand the distribution without any extra data processing.

In [13]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21

You'll notice that the first 10 characters represent the day (e.g. 2016-03-12). To understand the date range, we can extract just the date values, use `Series.value_counts()` to generate a distribution, and then sort by the index.

In [14]:
autos['date_crawled'].str[:10].head(10)

0     2016-03-24
1     2016-03-24
2     2016-03-14
3     2016-03-17
4     2016-03-31
5     2016-04-04
6     2016-04-01
8     2016-04-04
9     2016-03-17
10    2016-03-26
Name: date_crawled, dtype: object

We'll calculate the distribution of values in the `date_crawled`, `ad_created`, and `last_seen columns` (all string columns) as percentages.

In [15]:
print(autos['date_crawled'].isnull().value_counts())
print('\n')
print(autos['date_crawled'].value_counts(normalize=True, dropna=False))

False    360635
Name: date_crawled, dtype: int64


2016-03-24 14:49:47    0.000019
2016-03-26 22:57:31    0.000017
2016-03-19 21:49:56    0.000017
2016-03-05 14:25:23    0.000014
2016-04-02 21:54:36    0.000014
2016-03-28 10:36:47    0.000014
2016-03-26 10:51:07    0.000014
2016-04-01 16:52:05    0.000014
2016-03-16 13:47:44    0.000014
2016-04-01 19:25:23    0.000014
2016-04-01 15:56:46    0.000014
2016-03-09 13:50:56    0.000014
2016-03-07 17:36:19    0.000014
2016-04-04 22:38:11    0.000014
2016-03-22 10:42:10    0.000014
2016-03-31 18:50:16    0.000014
2016-03-14 16:51:53    0.000014
2016-03-05 14:44:30    0.000014
2016-04-03 16:49:06    0.000014
2016-03-29 22:50:49    0.000014
2016-04-03 13:41:21    0.000014
2016-03-21 14:50:20    0.000014
2016-03-28 10:48:11    0.000014
2016-04-02 22:54:55    0.000014
2016-04-02 14:50:21    0.000014
2016-03-23 13:50:19    0.000014
2016-03-22 14:50:05    0.000014
2016-03-31 17:57:07    0.000014
2016-03-16 17:43:49    0.000014
2016-03-27 12:50:22  

In [16]:
print(autos['ad_created'].isnull().value_counts())
print('\n')
autos['ad_created'].value_counts(normalize=True, dropna=False, ascending=False)

False    360635
Name: ad_created, dtype: int64




2016-04-03 00:00:00    0.039001
2016-04-04 00:00:00    0.037736
2016-03-20 00:00:00    0.036488
2016-03-12 00:00:00    0.036078
2016-03-21 00:00:00    0.035776
2016-03-14 00:00:00    0.035266
2016-03-28 00:00:00    0.035135
2016-04-02 00:00:00    0.034974
2016-03-07 00:00:00    0.034486
2016-03-19 00:00:00    0.034303
2016-03-09 00:00:00    0.034162
2016-03-29 00:00:00    0.034159
2016-04-01 00:00:00    0.033990
2016-03-30 00:00:00    0.033399
2016-03-08 00:00:00    0.033372
2016-03-15 00:00:00    0.033350
2016-03-25 00:00:00    0.032861
2016-03-11 00:00:00    0.032806
2016-03-10 00:00:00    0.032520
2016-03-22 00:00:00    0.032196
2016-03-26 00:00:00    0.032093
2016-03-23 00:00:00    0.031963
2016-03-31 00:00:00    0.031952
2016-03-17 00:00:00    0.031245
2016-03-16 00:00:00    0.030427
2016-03-27 00:00:00    0.030000
2016-03-24 00:00:00    0.029953
2016-03-05 00:00:00    0.022896
2016-03-13 00:00:00    0.017056
2016-03-06 00:00:00    0.015503
                         ...   
2016-01-

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [17]:
print(autos['last_seen'].isnull().value_counts())
print('\n')
autos['last_seen'].value_counts(normalize=True, dropna=False, ascending=False)

False    360635
Name: last_seen, dtype: int64




2016-04-07 06:45:59    0.000047
2016-04-07 00:45:17    0.000044
2016-04-07 06:45:12    0.000044
2016-04-07 07:17:35    0.000044
2016-04-07 08:45:09    0.000044
2016-04-06 01:15:23    0.000044
2016-04-07 09:15:43    0.000044
2016-04-06 09:44:33    0.000044
2016-04-06 12:15:45    0.000044
2016-04-06 13:45:54    0.000044
2016-04-06 03:15:56    0.000044
2016-04-07 05:44:30    0.000044
2016-04-07 13:17:48    0.000044
2016-04-06 05:45:43    0.000044
2016-04-07 07:46:26    0.000044
2016-04-06 10:18:03    0.000044
2016-04-07 07:44:31    0.000044
2016-04-06 02:45:52    0.000044
2016-04-07 09:44:27    0.000044
2016-04-07 00:46:04    0.000044
2016-04-06 17:46:58    0.000044
2016-04-06 04:17:20    0.000044
2016-04-06 04:15:59    0.000044
2016-04-06 09:46:00    0.000044
2016-04-06 08:46:30    0.000044
2016-04-06 09:17:58    0.000044
2016-04-07 08:16:54    0.000044
2016-04-06 01:16:21    0.000044
2016-04-07 06:16:14    0.000042
2016-04-07 07:15:24    0.000042
                         ...   
2016-03-

**Firstly**, there is no null values, which is significant.

**Secondly**, most of the ads are crawled and created at the end of the month.

**Thirdly**, ads are most last seen online by crawler at the beginning of April. It is unlikely that most of the cars are sold in those days. It is preferable that crawler's period is ending and it does not indicate car sales.

In [18]:
autos['registration_year'].describe()

count    360635.000000
mean       2004.433133
std          81.016977
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

*min* and *max* year look a little bit odd, respectively *1000* and *9999*. We will investigate further.

In [19]:
autos['registration_year'].value_counts()

2000    23122
1999    22031
2005    21725
2006    20005
2001    19744
2003    19606
2004    19473
2002    18820
2007    17506
1998    17226
2008    16035
2009    15498
1997    14029
2010    12267
2011    11996
1996    10410
2017     9996
2012     9359
2016     9216
1995     8999
2013     6121
2014     4769
1994     4646
2018     3814
1993     3298
2015     2919
1992     2896
1991     2657
1990     2503
1989     1259
        ...  
8200        1
7777        1
8500        1
6200        1
1602        1
7500        1
2900        1
1111        1
2066        1
1039        1
2290        1
6500        1
5300        1
9450        1
1001        1
1255        1
5600        1
3800        1
4800        1
7100        1
2222        1
3700        1
2800        1
2200        1
1925        1
3200        1
7800        1
1400        1
1911        1
4100        1
Name: registration_year, Length: 145, dtype: int64

Between 1990 and 1989, there is a hugh gap between them, so it's best to start from 1990 to make sure that the values are still relevant.

In [20]:
autos = autos[autos['registration_year'].between(1990,2019)]
autos['registration_year'].value_counts(normalize=True).sort_index(ascending=False)

2019    0.000048
2018    0.010875
2017    0.028503
2016    0.026279
2015    0.008323
2014    0.013598
2013    0.017454
2012    0.026686
2011    0.034206
2010    0.034978
2009    0.044191
2008    0.045722
2007    0.049917
2006    0.057043
2005    0.061947
2004    0.055526
2003    0.055905
2002    0.053664
2001    0.056298
2000    0.065930
1999    0.062820
1998    0.049118
1997    0.040003
1996    0.029683
1995    0.025660
1994    0.013248
1993    0.009404
1992    0.008258
1991    0.007576
1990    0.007137
Name: registration_year, dtype: float64

When working with data on cars, it's natural to explore variations across different car brands. We will use aggregation to understand the `brand` column.

In [21]:
print(autos['brand'].value_counts(normalize=True).head(20))
autos['brand'].value_counts(normalize=True).head(20).index

volkswagen       0.213169
bmw              0.109571
opel             0.108619
mercedes_benz    0.093823
audi             0.089845
ford             0.069247
renault          0.049335
peugeot          0.030693
fiat             0.025842
seat             0.019461
skoda            0.015825
mazda            0.015686
smart            0.014753
citroen          0.013963
nissan           0.013844
toyota           0.013008
hyundai          0.010237
mini             0.009487
volvo            0.008896
mitsubishi       0.008377
Name: brand, dtype: float64


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'smart', 'citroen',
       'nissan', 'toyota', 'hyundai', 'mini', 'volvo', 'mitsubishi'],
      dtype='object')

As shown, top 6 are the most distributed cars on eBay. We will only aggregate them to make our code become easier and relevant.

In [22]:
top_cars_mp = {}
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
for c in common_brands:
    mean_price = autos.loc[autos['brand'] == c, 'price'].mean()
    top_cars_mp[c] = int(mean_price)
    
print(top_cars_mp)

{'volkswagen': 5309, 'bmw': 8384, 'opel': 2912, 'mercedes_benz': 8370, 'audi': 9056, 'ford': 3502}


Top 6 cars' mean prices are pretty impressed. Although most people use Volkswagen as it's relatively cheap (top 1) or Opel (top 3), other people also buy expensive cars like BMW (top 2) or Mercedes Benz (top 4).

We will construct an another frame for mean price and mileage for top cars to be easily visible.

In [23]:
top_cars_ma = {}
for c in common_brands:
    mile_age = autos.loc[autos['brand'] == c, 'odometer_km'].mean()
    top_cars_ma[c] = int(mile_age)

In [35]:
mp_series = pd.Series(top_cars_mp)
ma_series = pd.Series(top_cars_ma)
ma_mp_frame = pd.DataFrame(ma_series, columns=['mileage'])
ma_mp_frame['mean_price'] = mp_series
ma_mp_frame

mileage  mean_price
volkswagen      129127        5309
bmw             133039        8384
opel            129499        2912
mercedes_benz   130805        8370
audi            129666        9056
ford            124704        3502

There is not many differences between Mileage and Mean Price of the top cars. Based on this, we can assume that, there are people who buy car economically, and those who don't.

*Till here I've done all the intructions from DQ*. 

From now on I will clean & analyze the data and analyze the dataset by myself and based on my reference.